### DISCUSSION TITLE
"sqlite3.OperationalError: database is locked" with multi-threaded app

### DISCUSSION THREAD

**Discussion by Chetan11-dev** [sqlalchemy/sqlalchemy#11524] [Usage Questions] [answered]
2024-06-22 06:44:14 Traceback (most recent call last):
2024-06-22 06:44:14   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
2024-06-22 06:44:14     self.dialect.do_execute(
2024-06-22 06:44:14   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
2024-06-22 06:44:14     cursor.execute(statement, parameters)
2024-06-22 06:44:14 sqlite3.OperationalError: database is locked
2024-06-22 06:44:14

**Comment by CaselIT** [comment]
hi,

Not much to go by here...
are you seeing this in a multi process application?

**Comment by Chetan11-dev** [comment]
Thanks I have created a library Botasaurus (https://github.com/omkarcloud/botasaurus/stargazers), 1000K+ Stars, it uses sqlalchemy in a multi threaded manner, do you think using a Lock Might Work?


```
from contextlib import contextmanager
import threading

engine = create_engine(db_url)
_Session = sessionmaker(bind=engine)
session_lock = threading.Lock()

@contextmanager
def Session():
    with session_lock:  # Acquire the lock
        with Session() as session:  # Create a new session
            yield session  # Yield the session to the caller
```
---
```
with Session() as session:
            # Delete tasks with is_sync=True and status as either pending or in progress
            session.query(Task).filter(
                and_(
                    Task.is_sync == True,
                    Task.status.in_([TaskStatus.PENDING, TaskStatus.IN_PROGRESS]),
                )
            )
```

or you know of a better solution?

**Reply by CaselIT** [reply]
That should not be necessary, since the important thing is not to share the same session in multiple threads.

The issue description is not clear though. Is the database locked error happening only after some long idle time?

Overall you are providing really very little information that can be used to suggest solutions.

**Comment by Chetan11-dev** [comment]
Basically, I have this code in db_setup.py:

```python
from .models import Base  # Assuming Task is one of your models

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine(db_url)
Session = sessionmaker(bind=engine)
Base.metadata.create_all(engine)
```

The Session is then imported in multiple files that need to access the database, like:

```python
from .db_setup import Session
from .models import Cache

def set_cache_data(key, data):
    with Session() as session:
        cache_entry = session.query(Cache).filter(Cache.key == key).first()
        if cache_entry:
            cache_entry.data = data
        else:
            cache_entry = Cache(key=key, data=data)
            session.add(cache_entry)
        session.commit()
```

Currently, the session is imported and used by:
- A thread that continuously polls the database every 1 second and starts scraping tasks
- A separate thread that saves results to the database when the scraper finishes (happens based on the time taken by the task, say every 1 minute)
- The Backend REST API that queries the database when a user makes requests, which occurs very infrequently (happens rarely when a user wants to check task status)

Also, the "Database is Locked" error does not occur until the database becomes big (600MB, 1000 scraping tasks completed). After that, it starts happening more frequently.

**Reply by gordthompson** [reply]
Have you seen this? https://sqlite.org/faq.html#q5

**Reply by CaselIT** [reply]
It seems related to it. I don't think changing the pool implementation will fix this, but you could try using a nullpool instead of the default queuepool.

Also worth trying increasing this timeout https://docs.python.org/3/library/sqlite3.html#sqlite3.connect

**Comment by Chetan11-dev** [comment]
Thanks @CaselIT and @gordthompson, I will try to use timeouts and poolclass and will let you know if this resolved it,

from sqlalchemy import NullPool
engine = create_engine(db_url, 
                       pool_timeout=60, poolclass=NullPool,)

**Reply by CaselIT** [reply]
I believe that to change the sqlite3 connection timeout you have to use `connect_args={'timeout': 60}`

**Reply by Chetan11-dev** [reply]
Thanks!

**Reply by Chetan11-dev** [reply, marked as answer]
Thanks, the error occurred, as scraping results were too large, make sqlite inflate too many 100s of MBs, causing such errors. 
We solved it by saving results separately as json and ndjson files.

I have send $50 as a thank you, I know it is not much considering your effort, but this the best I can at the moment.
Good Luck

**Reply by CaselIT** [reply]
> I have send $50 as a thank you, I know it is not much considering your effort, but this the best I can at the moment.

Thanks, that's more than enough! This is not a paid support forum, people are free to donate or not as they please

**Reply by Chetan11-dev** [reply]
Great, Keep doing the Good Work!
